### PR TITLE
Feature/44 external social login

### LINKED ISSUE #76: Authentication of user on server
Every login request from user gets authenticated on server side, using tokenId and emailId from Google

### Story

#44 
#47 

### Estimate
6h

### LINKED ISSUE #87: Creation of /login REST API
/login api called from client whenever any user clock on login button

##Estimate

4h

##User Story

This task belongs to user story #44

### LINKED ISSUE #92: Create/Update user in database
User details have been saved in the database for first-time login. Later, status of user has been updated in database to ONLINE on every login request

##Estimate

2h

##User Story

This task belongs to user story #44

### PR DESCRIPTION
**Pull request by nehasingh1491**
Addition of login functionality on the server-side

Closing #76 #87 #92

### CODE REVIEW

File: api/openapi.yaml [hunk 1]
```diff
@@ -129,19 +129,23 @@ paths:
...
-          $ref: "#/components/responses/400"
+          $ref: '#/components/responses/400'
+        403:
+          $ref: '#/components/responses/403'
         401:
           $ref: "#/components/responses/401"
-        403:
-          $ref: "#/components/responses/403"
+        404:
+          $ref: "#/components/responses/404"
```
Comment (KseniaBel) [inline api/openapi.yaml:148]: What is the purpose of 404 in this usecase?

File: src/main/java/ch/uzh/ifi/hase/soprafs21/service/UserService.java [hunk 2]
```diff
@@ -47,5 +59,70 @@ public User createUser(User newUser) {
...
     }
+
+    /**
+     *
+     * @param tokenId
+     * @param emailId
+     * @return
+     * @throws ResponseStatusException
+     */
+    public Boolean verifyEmailIdForToken(GoogleIdToken tokenId, String emailId) throws ResponseStatusException{
```
Comment (KseniaBel) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/service/UserService.java:93]: What is the purpose of this method?

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java [hunk 3]
```diff
@@ -12,13 +20,40 @@
...
 
+    private final UserService userService;
+
     @Override
     public Optional<NativeWebRequest> getRequest() {
         return Optional.ofNullable(request);
     }
 
+    @Override
+    public ResponseEntity<Boolean> usersLoginPost(@RequestBody UserLoginDto userLoginDto) throws GeneralSecurityException, IOException {
```
Comment (KseniaBel) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java:36]: I understand what you are doing here, but I have one question: how will we apply security check for all other rest APIs?

File: api/openapi.yaml (hunk 1 above)
Comment (nehasingh1491) [reply api/openapi.yaml:148]: If token passed is not present on google server, or it is null.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/service/UserService.java (hunk 2 above)
Comment (nehasingh1491) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/service/UserService.java:93]: It's just for second-level check in the application that the provided token is valid for the user accessing endpoint. Also helpful for api requests other than app UI like postman, etc.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java (hunk 3 above)
Comment (nehasingh1491) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java:36]: We can add tokenId in Dto's and call helper method to authenticate first for all rest API's.

**Pull request merged**